src/neuralnet/neuralnet.cc

```diff
@@ -99,11 +107,89 @@ NeuralNet* NeuralNet::Create(const NetProto& net_conf, Phase phase,
     param->set_name(name);
     param->set_share_from(from);
   }
+
   LOG(INFO) << "NeuralNet config is\n" << conf.DebugString();
   // TODO(wangwei) create net based on net type, e.g., directed, undirected, etc
   return new NeuralNet(conf, npartitions);
 }
 
+NetProto NeuralNet::Unrolling(const NetProto& net_conf) {
+	// Step 1: Unroll each layer & set parameter sharing
+	NetProto conf;
+	conf.CopyFrom(net_conf);
+	conf.Clear();
+
+	std::vector<std::vector<int>> layer_groups;
+	std::unordered_map<string,int> layer_names;
+	for (int index = 0; index < net_conf.layer_size(); index ++) {
+		const LayerProto& layer = net_conf.layer(index);
+		layer_names[layer.name()] = index; // layer_name -> index
+
+		std::vector<int> layer_group;
+		for (int i = 0; i < layer.unroll_len(); i ++) { // unroll
+			LayerProto* layer_conf = conf.add_layer();
+			layer_conf->CopyFrom(layer); // create a new layer conf
+			if (layer.unroll_len() > 1) {
+				// update layer names
+				std::stringstream sstm;
+				sstm << layer_conf->name() << "_" << i;
+				layer_conf->set_name(sstm.str());
+				// update layer parameter sharing
+				for (int j = 0; j < layer_conf->param_size(); j ++) {
+					ParamProto* param = layer_conf->mutable_param(j);
+					if (i == 0) continue; // no need to rename parameters in the i-th unrolled layer
+					if (!param->has_share_from() || param->share_from() == "") {// not shared from others
+						param->set_share_from(param->name());
+					}
+					std::stringstream sstm1;
+					sstm1 << param->name() << "_" << i;
+					param->set_name(sstm1.str());
+				}
+			}
+			// clear unrolling related fields
+			layer_conf->clear_unroll_len();
+			layer_conf->clear_unroll_conn_type();
+			layer_conf->clear_shift();
+			layer_conf->clear_srclayers();
+
+			layer_group.push_back(conf.layer_size() - 1);
+		}
+		layer_groups.push_back(layer_group);
+	}
+	// Step 2: Connect unrolled layers by setting `srclayers`
```

Comment:
thanks. I can help update the code if you add me as your collaborator. but my network is not stable these two days..